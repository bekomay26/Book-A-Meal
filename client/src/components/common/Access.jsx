import 'babel-polyfill';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import jwt from 'jsonwebtoken';
import moment from 'moment';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import { logout } from '../../actions/authActions';

export class Access extends Component {
  constructor(props) {
    super(props);

    this.logoutNow = this.logoutNow.bind(this);
    this.tokenExpired = this.tokenExpired.bind(this);
  }

  /**
   * method to handle logout.
   * @param {object} event
   */
  logoutNow(event) {
    event.preventDefault();
    this.props.logout();
  }

  /**
   * method to verify if user token has expired
   */
  tokenExpired() {
    if (localStorage.jwtToken) {
      const token = localStorage.jwtToken;
      const { exp } = jwt.decode(token);
      if (moment(Math.floor(Date.now() * 0.001)).isAfter(exp)) {
        return false;
      }
      return true;
    }
    return false;
  }

  render() {
    /**
     * Check if user token is still valid
     */
    if (!this.tokenExpired()) {
      const event = {
        preventDefault: () => Promise.resolve(),
      };
      this.logoutNow(event);
      return (
        <Redirect to="/login" />
      );
    }

    /**
     * Redirects unauthenticated users
     */
    if (!this.props.allowAnonymous) {
      if (!this.props.isAuthenticated) {
        return (
          <Redirect to="/login" />
        );
      }
    }

    /**
     * Restricts customer from accessing page if customer is not allowed
     */
    if (!this.props.allowCustomer) {
      if (!this.props.isCaterer && this.props.isAuthenticated) {
        return (
          <Redirect to="/unauthorized" />
        );
      }
    }

    /**
     * Handles output rendered by the component.
     * @returns {jsx} form of the component.
     */
    return (
      <div isAuthenticated={this.props.isAuthenticated} isCaterer={this.props.isCaterer} />
    );
  }
}

// proptypes
/* eslint-disable react/forbid-prop-types */
Access.propTypes = {
  isAuthenticated: PropTypes.bool.isRequired,
  isCaterer: PropTypes.bool.isRequired,
  logout: PropTypes.func.isRequired,
  allowAnonymous: PropTypes.bool,
  allowCustomer: PropTypes.bool,
};

// default props
Access.defaultProps = {
  allowAnonymous: false,
  allowCustomer: false,
};

/**
 * @desc maps state to props;
 * @param {*} state store state
 * @returns {*} store state
 */
const mapStateToProps = state => (
  {
    user: state.authReducer.user,
    isAuthenticated: state.authReducer.isAuthenticated,
    isCaterer: state.authReducer.isCaterer,
  }
);

/**
 * @desc maps dispatch to props;
 * @param {*} dispatch dispatch
 * @returns {*} action to be dispatched
 */
const mapDispatchToProps = dispatch => bindActionCreators({ logout }, dispatch);


export default connect(mapStateToProps, mapDispatchToProps)(Access);
