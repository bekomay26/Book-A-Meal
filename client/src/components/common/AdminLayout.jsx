import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Layout, Menu, Breadcrumb } from 'antd';
import { NavLink, Redirect } from 'react-router-dom';
import UserNavBar from './UserNavBar';
import AdminNavBar from './AdminNavBar';
import '../../assets/styles/adminlayout.css';

export class AdminLayout extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      isDesktop: false,
    };
    this.updatePredicate = this.updatePredicate.bind(this);
  }
  componentDidMount() {
    this.updatePredicate();
    window.addEventListener('resize', this.updatePredicate);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.updatePredicate);
  }

  updatePredicate() {
    this.setState({ isDesktop: window.innerWidth > 960 });
  }

  render() {
    const { isDesktop } = this.state;
    const { content } = this.props;
    const { SubMenu } = Menu;
    const { Content, Sider } = Layout;
    const desktopPage = (
      <Layout>
        <AdminNavBar />
        <Layout className="sidelayout">
          <Sider width={200} style={{ background: '#fff' }}>
            <Menu mode="inline" defaultSelectedKeys={['1']} defaultOpenKeys={['sub1']} style={{ height: '100%', borderRight: 0 }} >
              <Menu.Item key="sub1">
                <span><img src="https://png.icons8.com/ios/80/ffffff/combo-chart.png" alt="analytics" />Dashboard</span>
              </Menu.Item>
              <SubMenu key="sub2" title={<span><img src="https://png.icons8.com/wired/80/ffffff/hamburger.png" alt="meals" />Meals</span>}>
                <Menu.Item key="5">
                  <NavLink className="admin-side navlink" to="/meals">View Meals</NavLink>
                </Menu.Item>
                <Menu.Item key="6">Add Meal</Menu.Item>
                <Menu.Item key="7">Add Extra</Menu.Item>
              </SubMenu>
              <Menu.Item key="sub3">
                <span>
                  <img src="https://png.icons8.com/ios/80/ffffff/checked-truck.png" alt="orders" />
                  <NavLink className="admin-side navlink" to="/adorders">Orders</NavLink>
                </span>
              </Menu.Item>
              <SubMenu key="sub4" title={<span><img src="https://png.icons8.com/ios/80/ffffff/restaurant-menu.png" alt="menu" />Menu</span>}>
                <Menu.Item key="9">
                  <NavLink className="admin-side navlink" to="/setmenu"><div onClick={() => { this.props.redirectTo(3); }}> Set Menu</div></NavLink>
                </Menu.Item>
                <Menu.Item key="10" onClick={<Redirect push to="/setmenu" />} >Edit Menu</Menu.Item>
                <Menu.Item key="11" onClick={(event) => { this.props.redirectTo(event, 3); }}>Menu History</Menu.Item>
              </SubMenu>
            </Menu>
          </Sider>
          <Layout style={{ padding: '0 24px 24px' }}>
            <Breadcrumb style={{ margin: '16px 0' }}>
              <Breadcrumb.Item>Home</Breadcrumb.Item>
              <Breadcrumb.Item>List</Breadcrumb.Item>
              <Breadcrumb.Item>App</Breadcrumb.Item>
            </Breadcrumb>
            <Content style={{
              background: '#fff',
              padding: 24,
              margin: 0,
              minHeight: 650,
              maxHeight: 650,
              overflow: 'auto',
            }}
            >
              {content}
            </Content>
          </Layout>
        </Layout>
      </Layout>);
    const mobilePage = (
      <div>
        <AdminNavBar />
        { content }
      </div>
    );
    return (
      <div>
        {/* {desktopPage} */}
        { isDesktop ? desktopPage : mobilePage }
      </div>
    );
  }
}
AdminLayout.propTypes = {

};

/**
 * @desc maps state to props;
 * @param {*} state store state
 * @returns {*} store state
 */
const mapStateToProps = state => ({
  // isAuthenticated: state.authReducer.isAuthenticated,
  // isCaterer: state.authReducer.isCaterer,
  // userName: state.authReducer.userName,
});

export default connect(mapStateToProps)(AdminLayout);
