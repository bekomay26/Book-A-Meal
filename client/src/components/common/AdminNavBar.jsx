import React, { Component } from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Icon, Drawer } from 'antd';
import { bindActionCreators } from 'redux';
import { logout } from '../../actions/authActions';
import '../../assets/styles/navbar.css';

export class AdminNavBar extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      visibleSideBar: false,
    };
    this.openSideNav = this.openSideNav.bind(this);
    this.handleSidebarHide = this.handleSidebarHide.bind(this);
  }

  openSideNav() { this.setState({ visibleSideBar: true }); }

  handleSidebarHide() { this.setState({ visibleSideBar: false }); }
  render() {
    const { userName } = this.props;
    const { isAuthenticated } = this.props;
    const { visibleSideBar } = this.state;
    let navlinkVisibility = 'show';
    let signedinVisibility = 'hide';
    if (isAuthenticated) {
      navlinkVisibility = 'hide';
      signedinVisibility = 'show';
    }
    return (
      <div>
        <div className="admin-top topnav row" id="myTopnav">
          <button className="icon" onClick={this.openSideNav}>
            <i className="fa fa-bars" />
          </button>
          <div className="col-6">
            <NavLink className="navlogo col-5" to="/">BOOK-A-MEAL</NavLink>
          </div>
          <nav className="col-6 navgroup">
            <li className={`col-5 navlink admin-nav ${signedinVisibility}`}>{<span><Icon type="user" />{userName}</span>}</li>
            <NavLink className={`col-3 navlink out admin-nav ${signedinVisibility}`} activeClassName="active" to="/login" onClick={(e) => { e.preventDefault(); this.props.logout(); }}>logout</NavLink>
          </nav>
        </div>
        <Drawer
          className="admin-drawer"
          title="Book-A-Meal"
          width={400}
          placement="left"
          maskClosable={false}
          onClose={this.handleSidebarHide}
          visible={visibleSideBar}
        >
          <NavLink className="col-3 show navlink" to="/meals" activeClassName="active">Meals</NavLink>
          <NavLink className="col-3 show navlink" to="/setmenu" activeClassName="active">Menu</NavLink>
          <NavLink className="col-3 show navlink" to="/adorders" activeClassName="active">Orders</NavLink>
          <NavLink className={`col-3 navlink ${navlinkVisibility}`} to="/login" activeClassName="active">SignIn</NavLink>
          <NavLink className={`col-3 navlink out ${signedinVisibility}`} activeClassName="active" to="/login" onClick={(e) => { e.preventDefault(); this.props.logout(); }}>logout</NavLink>
        </Drawer>
      </div>
    );
  }
}

AdminNavBar.defaultProps = {
  userName: undefined,
  isAuthenticated: false,
};

AdminNavBar.propTypes = {
  userName: PropTypes.string,
  isAuthenticated: PropTypes.bool,
  logout: PropTypes.func.isRequired,
};

/**
 * @desc maps dispatch to props;
 * @param {*} dispatch dispatch
 * @returns {*} action to be dispatched
 */
const mapDispatchToProps = dispatch => bindActionCreators({ logout }, dispatch);

/**
 * @desc maps state to props;
 * @param {*} state store state
 * @returns {*} store state
 */
const mapStateToProps = state => ({
  isAuthenticated: state.authReducer.isAuthenticated,
  isCaterer: state.authReducer.isCaterer,
  userName: state.authReducer.userName,
});

export default connect(mapStateToProps, mapDispatchToProps)(AdminNavBar);
