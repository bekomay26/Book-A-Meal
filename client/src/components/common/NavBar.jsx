// import React from 'react';
// import { Link, NavLink } from 'react-router-dom';

// const NavBar = () => (
//   <div>
//     <div className="back-image" />
//     <div className="top row">
//       <Link to="/h" className="col-8 logo"><h2>Book-A-Meal</h2></Link>
//       <nav className="menu col-4">
//         <li className="col-3"><NavLink to="/menu" activeClassName="active">Menu</NavLink></li>
//         <li className="col-3"><NavLink to="/cart" activeClassName="active">Order</NavLink></li>
//         <li className="col-3"><NavLink to="/login" activeClassName="active">SignIn</NavLink></li>
//         <li className="col-3"><NavLink to="/signup" activeClassName="active">Signup</NavLink></li>
//       </nav>
//     </div>
//   </div>
// );
// export default NavBar;


import React, { Component } from 'react';
import { Link, NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

class NavBar extends Component {

  render() {
    let navlinkVisibility = 'show';
    let signedinVisibility = 'hide';
    if (this.props.isAuthenticated) {
      navlinkVisibility = 'hide';
      signedinVisibility = 'show';
    }
    return (<div>
    <div className="back-image" />
    <div className="top row">
      <Link to="/" className="col-8 logo"><h2>Book-A-Meal</h2></Link>
      <nav className="menu col-4">
        <li className={`col-3 ${signedinVisibility}`}>{this.props.userName}</li>
        <li className="col-3"><NavLink to="/menu" activeClassName="active">Menu</NavLink></li>
        <li className="col-3"><NavLink to="/myorders" activeClassName="active">Order</NavLink></li>
        <li className={`col-3 ${navlinkVisibility}`}><NavLink to="/login" activeClassName="active">Login</NavLink></li>
        <li className={`col-3 ${signedinVisibility}`}><NavLink to="/login" onClick={(e) => { e.preventDefault(); logout(); }}>Logout</NavLink></li>
      </nav>
    </div>
  </div>)
  }
};

NavBar.defaultProps = {
  userName: undefined,
  isAuthenticated: false,
};

NavBar.propTypes = {
  userName: PropTypes.string,
  isAuthenticated: PropTypes.bool,
};


/**
 * @desc maps state to props;
 * @param {*} state store state
 * @returns {*} store state
 */
const mapStateToProps = state => ({
  isAuthenticated: state.authReducer.isAuthenticated,
  isCaterer: state.authReducer.isCaterer,
  userName: state.authReducer.userName,
});

export default connect(mapStateToProps)(NavBar);
