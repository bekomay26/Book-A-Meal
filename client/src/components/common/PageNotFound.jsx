import React, { Fragment } from 'react';
import { NavLink } from 'react-router-dom';
import '../../assets/styles/otherpages.css';

const PageNotFound = () => (
  <Fragment>
    <div className="wrong-page-img2" />
    <div className="overlay" />
    <div className="wrong-page-content">
      <h1>404</h1>
      <h3>Whoops...!!!</h3>
      <h3>We could not find the meal you were looking for</h3>
      <br />
      <NavLink className="wrong-page-btn" to="/menu">See the menu</NavLink>
    </div>
  </Fragment>
);

export default PageNotFound;
