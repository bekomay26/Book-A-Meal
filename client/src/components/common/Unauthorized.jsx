import React, { Fragment } from 'react';
import { NavLink } from 'react-router-dom';
import '../../assets/styles/otherpages.css';

const Unauthorized = () => (
  <Fragment>
    <div className="wrong-page-img" />
    <div className="overlay" />
    <div className="wrong-page-content">
      <h1>403</h1>
      {/* <h2>Unauthorized</h2> */}
      <h3>Whoops...</h3>
      <h3>It seems you are not allowed in the kitchen!!!</h3>
      <br />
      <NavLink className="wrong-page-btn" to="/">Go home</NavLink>
    </div>
  </Fragment>
);

export default Unauthorized;
