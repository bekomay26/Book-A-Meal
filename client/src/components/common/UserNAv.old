// import React from 'react';
// import { Link, NavLink } from 'react-router-dom';

// const NavBar = () => (
//   <div>
//     <div className="back-image" />
//     <div className="top row">
//       <Link to="/h" className="col-8 logo"><h2>Book-A-Meal</h2></Link>
//       <nav className="menu col-4">
//         <li className="col-3"><NavLink to="/menu" activeClassName="active">Menu</NavLink></li>
//         <li className="col-3"><NavLink to="/cart" activeClassName="active">Order</NavLink></li>
//         <li className="col-3"><NavLink to="/login" activeClassName="active">SignIn</NavLink></li>
//         <li className="col-3"><NavLink to="/signup" activeClassName="active">Signup</NavLink></li>
//       </nav>
//     </div>
//   </div>
// );
// export default NavBar;


import React, { Component } from 'react';
import { Link, NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import styles from '../../assets/styles/navbar.css';
import { Icon } from 'antd';
import { Sidebar, Segment, Menu, Header, Image } from 'semantic-ui-react';

const UserNavBar = ({ logout, isAuthenticated, userName, visibleSideBar }) => {
  let navlinkVisibility = 'show';
  let signedinVisibility = 'hide';
  if (isAuthenticated) {
    navlinkVisibility = 'hide';
    signedinVisibility = 'show';
  }
  const openNav = () => this.setState({ visibleSideBar: true });

  const handleSidebarHide = () => this.setState({ visibleSideBar: false });
  // const openNav = () => {
  //   visibleSideBar = true;
  //   return visibleSideBar;
  // };
  // const handleSidebarHide = () => {
  //   visibleSideBar = false;
  //   return visibleSideBar;
  // };
  return (
    <div>
      <div className="topnav row" id="myTopnav">
        <button className="icon" onClick={openNav()}>
          <i className="fa fa-bars" />
          {/* <Icon type="menu-unfold" /> */}
        </button>
        <div className="col-6">
          <a href="#home" className="navlogo col-5">BOOK-A-MEAL</a>
        </div>
        <nav className="col-6 navgroup">
          <li className={`col-3 navlink ${signedinVisibility}`}>{userName}</li>
          <NavLink className="col-3 show navlink" to="/menu" activeClassName="active">Menu</NavLink>
          <NavLink className="col-3 show navlink" to="/myorders" activeClassName="active">Order</NavLink>
          <NavLink className={`col-3 navlink ${navlinkVisibility}`} to="/auth" activeClassName="active">SignIn</NavLink>
          {/* <NavLink className={`col-3 navlink ${navlinkVisibility}`} to="/signup" activeClassName="active">SignUp</NavLink> */}
          <NavLink className={`col-3 navlink out ${signedinVisibility}`} activeClassName="active" to="/login" onClick={(e) => { e.preventDefault(); logout(); }}>logout</NavLink>
        </nav>
      </div>
      <Sidebar.Pushable as={Segment}>
        <Sidebar
          as={Menu}
          animation="overlay"
          icon='labeled'
          inverted
          onHide={handleSidebarHide()}
          vertical
          visible={visibleSideBar}
          width='thin'
        >
          <Menu.Item as='a'>
            <Icon name='home' />
            Home
          </Menu.Item>
          <Menu.Item as='a'>
            <Icon name='gamepad' />
            Games
          </Menu.Item>
          <Menu.Item as='a'>
            <Icon name='camera' />
            Channels
          </Menu.Item>
        </Sidebar>

        <Sidebar.Pusher dimmed={visibleSideBar}>
          <Segment basic>
            <Header as='h3'>Application Content</Header>
            <Image src='https://react.semantic-ui.com/images/wireframe/paragraph.png' />
          </Segment>
        </Sidebar.Pusher>
      </Sidebar.Pushable>
    </div>
  );
};

UserNavBar.defaultProps = {
  userName: undefined,
  isAuthenticated: false,
  isCaterer: false,
};

UserNavBar.propTypes = {
  userName: PropTypes.string,
  isAuthenticated: PropTypes.bool,
  isCaterer: PropTypes.bool,
  logout: PropTypes.func.isRequired,
};


/**
 * @desc maps state to props;
 * @param {*} state store state
 * @returns {*} store state
 */
const mapStateToProps = state => ({
  isAuthenticated: state.authReducer.isAuthenticated,
  isCaterer: state.authReducer.isCaterer,
  userName: state.authReducer.userName,
});

export default connect(mapStateToProps)(UserNavBar);
