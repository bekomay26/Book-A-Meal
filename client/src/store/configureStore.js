import { createStore, applyMiddleware } from 'redux';
import reduxImmutableStateInvariant from 'redux-immutable-state-invariant';
import thunk from 'redux-thunk';
import rootReducer from '../reducers/index';


const configureStore = initialState =>
  createStore(
    rootReducer,
    initialState,
    applyMiddleware(thunk, reduxImmutableStateInvariant()),
  );

export default configureStore;
