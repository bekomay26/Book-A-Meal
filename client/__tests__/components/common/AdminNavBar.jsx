import React from 'react';
import { shallow, mount, render } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { AdminNavBar } from '../../../src/components/common/AdminNavBar';

// import jsdom from 'jsdom';
// const doc = jsdom.jsdom('<!doctype html><html><body></body></html>')
// global.document = doc
// global.window = doc.defaultView
const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const store = mockStore({
  authReducer: {
    isAuthenticated: false,
    isCaterer: false,
    errors: {},
  },
});

function setup() {
  return shallow(<AdminNavBar />, reactrouter.get());
}

describe('The Admin Nav Bar Component', () => {
  it('should render admin nav bar component correctly', () => {
    const wrapper = setup();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
