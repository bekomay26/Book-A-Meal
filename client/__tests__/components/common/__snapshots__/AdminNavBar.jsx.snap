// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Admin Nav Bar Component should render admin nav bar component correctly 1`] = `
<div>
  <div
    className="admin-top topnav row"
    id="myTopnav"
  >
    <button
      className="icon"
      onClick={[Function]}
    >
      <i
        className="fa fa-bars"
      />
    </button>
    <div
      className="col-6"
    >
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="navlogo col-5"
        to="/"
      >
        BOOK-A-MEAL
      </NavLink>
    </div>
    <nav
      className="col-6 navgroup"
    >
      <li
        className="col-5 navlink admin-nav hide"
      >
        <span>
          <Icon
            type="user"
          />
        </span>
      </li>
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="col-3 navlink out admin-nav hide"
        onClick={[Function]}
        to="/login"
      >
        logout
      </NavLink>
    </nav>
  </div>
  <Drawer
    className="admin-drawer"
    closable={true}
    height={256}
    level={null}
    maskClosable={false}
    onClose={[Function]}
    placement="left"
    prefixCls="ant-drawer"
    title="Book-A-Meal"
    visible={false}
    width={400}
  >
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="col-3 show navlink"
      to="/meals"
    >
      Meals
    </NavLink>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="col-3 show navlink"
      to="/setmenu"
    >
      Menu
    </NavLink>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="col-3 show navlink"
      to="/adorders"
    >
      Orders
    </NavLink>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="col-3 navlink show"
      to="/login"
    >
      SignIn
    </NavLink>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="col-3 navlink out hide"
      onClick={[Function]}
      to="/login"
    >
      logout
    </NavLink>
  </Drawer>
</div>
`;
